### ISSUE TITLE
Message code resolving in Web Flow seems broken [SWF-977]

### ISSUE THREAD

**Issue by spring-operator** [spring-projects/spring-webflow#1676]
**[Grzegorz Borkowski](https://jira.spring.io/secure/ViewProfile.jspa?name=grzegorzborkowski)** opened **[SWF-977](https://jira.spring.io/browse/SWF-977?redirect=false)** and commented

I have the Portlet application based on SWF and MVC.
I have the form displaying the order, which have table for displaying order items. In each row there is an editable field "Quantity" so that use can edit the number of items ordered.
The 'order' model object contains 'orderItems' list with items; so the input fields in the table are binded to this expression:
first row:      order.orderItems[0].quantity
second row: order.orderItems[1].quantity
etc.
Quantity is a number. If user enters the non-number string, the binding error is raised and message is displayed: "typeMismatch on order.orderItems[0].quantity". This is obviously too obscure for user, so I want to replace it with some nice message. Here begins the problem.

1. I searched across SWF user manual, find no word about replacing the messages. I believe that standard procedure is used, i.e. in messages.properties file I specify the messages, but what are the keys (codes) to use? No idea. It is not documented at all. I search also across forums, but also no hint found.
2. We use Spring MVC, so perhaps something in Spring Framework manual. I found only the section on binding referring to DefaultMessageCodesResolver. According to the documentation of DefaultMessageCodesResolver, I should expect following codes to be used:
   * 1. try "typeMismatch.order.orderItems[0].quantity"
   * 2. try "typeMismatch.order.orderItems.quantity"
   * 3. try "typeMismatch.orderItems[0].quantity"
   * 4. try "typeMismatch.orderItems.quantity"
   * 5. try "typeMismatch.quantity"
   * 6. try "typeMismatch.int"
   * 7. try "typeMismatch"
        I tried all of them, and found out that only the last one ("typeMismatch" ) works. It's sad, because it is too general, not specific to this order object or this screen.
        So how should I specify the code for this situation?
3. After debugging the SWF code, I found that the method AbstractMvcView.createMessageResolver() is called and it creates two codes:
   * 1. "order.orderItems[0].quantity.typeMismatch"
   * 2. "typeMismatch"
        Again, first one makes no sense in case of list (I cannot list all possible indices in messages file: "order.orderItems[0].quantity.typeMismatch", "order.orderItems[1].quantity.typeMismatch", "order.orderItems[2].quantity.typeMismatch", ... - it makes no sense). The second one is too general (actually, I used it, because I had no choice).

So it looks like there is no simple way to specify message in such a case.

To sum it up:
1. The message codes used by SWF are not documented. One have to debug the code to find it.
2. The mechanism of creating those codes is not documented too. Does it mean that SWF doesn't use the DefaultMessageCodesResolver?
3. If SWF uses its own mechanism instead of DefaultMessageCodesResolver, then this mechanism is very poor, as you see: there is no way to specify messages for collections binding errors; also, there is no way of specifying messages for specific types (e.g. no code "typeMismatch.int").
4. DefaultMessageCodesResolver uses codes patterns "errorType + object" ("typeMismatch.x.x.x"). The SWF mechanism uses opposite: "object + errorType" ("x.x.x.typeMismatch"). This is misleading.

---

**Affects:** 2.0.5
